experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts

```diff
@@ -0,0 +1,193 @@
+						context,
+						typeSet,
+						node as ContextuallyTypedNodeData,
+					);
+					assert(
+						possibleTypes.length !== 0,
+						"data is incompatible with all types allowed by the schema",
+					);
+					assert(
+						possibleTypes.length === 1,
+						"data is compatible with more than one type allowed by the schema",
+					);
+					return possibleTypes[0];
+				}
+			}
+		},
+		keysFromNode: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
+						// A list with no children is considered to not have any fields.
+						// If it has children, return the special symbol indicating a list.
+						return node.length === 0 ? [] : [EmptyKey];
+					} else if (node instanceof Map) {
+						const unfilteredKeys = Array.from(node.keys()) as FieldKey[];
+						// Map proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the entry.
+						const filteredKeys = unfilteredKeys.filter(
+							(key) => node.get(key) !== undefined,
+						);
+						return filteredKeys;
+					} else {
+						// Assume record-like object
+						const objectedNode = node as Record<FieldKey, unknown>;
+
+						const unfilteredKeys = Object.keys(node) as FieldKey[];
+						// Object proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the key/value pair.
+						const filteredKeys = unfilteredKeys.filter((key) => {
+							const value = objectedNode[key];
+							return value !== undefined;
+						});
+						return filteredKeys;
+					}
+				default:
+					return [];
+			}
+		},
+		getFieldFromNode: (node: TNode, key: FieldKey): readonly TNode[] => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+
+			// Object.prototype.hasOwnProperty can return true for strings (ex: with key "0"), so we have to filter them out.
+			// Rather than just special casing strings, we can handle them with an early return for all primitives.
+			if (typeof node !== "object") {
+				return [];
+			}
+
+			if (node === null) {
+				return [];
+			}
+
+			if (Array.isArray(node)) {
+				return key === EmptyKey || key === rootFieldKey ? node : [];
+			}
+
+			if (node instanceof Map) {
+				const value = (node as Map<FieldKey, TNode>).get(key);
+				return value === undefined ? [] : [value];
+			}
+
+			if (Object.prototype.hasOwnProperty.call(node, key)) {
+				const field = (node as Record<FieldKey, TNode>)[key];
+				return field === undefined ? [] : [field];
```

Comment:
Daniel is supposed to fix that we report proxies as having keys when the field is undefined, so this check should really be an assert (or t lest have a TODO to become one when thats fixed)

Reply:
But `node` is not a proxy. It is just a POJO.